File: stepik-union/src/main/java/org/stepik/plugin/projectWizard/ui/CourseListBox.java
package org.stepik.plugin.projectWizard.ui;

import com.intellij.openapi.project.Project;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.stepik.api.objects.courses.Course;

import javax.swing.*;
import java.awt.event.ItemEvent;

/**
 * @author meanmail
 */
public class CourseListBox extends JComboBox<Course> {
    private final CourseListModel courseListModel;
    private ProjectSetting target;

    public CourseListBox() {
        super();
        courseListModel = new CourseListModel();
        setModel(courseListModel);
        CourseListBoxEditor courseEditor = new CourseListBoxEditor();
        courseEditor.setModel(courseListModel);
        courseEditor.setOwner(this);
        setEditor(courseEditor);
    }

    void refresh(@NotNull Project project) {
        courseListModel.update(project);
    }

    void setTarget(@Nullable ProjectSetting target) {
        this.target = target;
    }

    @NotNull
    @Override
    public Course getSelectedItem() {
        return courseListModel.getSelectedItem();
    }

    @Override
    protected void fireItemStateChanged(@NotNull ItemEvent e) {
        super.fireItemStateChanged(e);

        if (target != null && e.getStateChange() == ItemEvent.SELECTED) {
            target.selectedCourse((Course) e.getItem());
        }
    }
}

File: stepik-union/src/main/kotlin/org/stepik/plugin/projectWizard/ui/CourseListBox.kt
package org.stepik.plugin.projectWizard.ui

import org.stepik.api.objects.StudyObject
import org.stepik.core.SupportedLanguages
import java.awt.event.ItemEvent
import javax.swing.JComboBox


class CourseListBox : JComboBox<StudyObject>() {
    private val courseListModel: CourseListModel = CourseListModel()
    internal var target: ProjectSetting? = null
    private var programmingLanguage = SupportedLanguages.INVALID

    init {
        model = courseListModel
        val courseEditor = CourseListBoxEditor()
        courseEditor.model = courseListModel
        courseEditor.owner = this
        setEditor(courseEditor)
    }

    internal fun refresh(programmingLanguage: SupportedLanguages) {
        this.programmingLanguage = programmingLanguage
        courseListModel.update(programmingLanguage)
    }

    internal fun refresh() {
        courseListModel.update(programmingLanguage)
    }

    override fun getSelectedItem(): StudyObject {
        return courseListModel.selectedItem
    }

    override fun fireItemStateChanged(e: ItemEvent) {
        super.fireItemStateChanged(e)

        if (e.stateChange == ItemEvent.SELECTED) {
            target?.selectedStudyNode(e.item as StudyObject)
        }
    }
}
